### PR TITLE
Add ability to associate metadata tags with Pod UID

### PR DESCRIPTION
**Pull request by pmatyjasek-sumo**
Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

**Description:** <Describe what has changed.>

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: processor/k8sprocessor/pod_association.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sprocessor
+
+import (
+	"context"
+	"net"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/kube"
```
Comment (pmm-sumo) [inline processor/k8sprocessor/pod_association.go:21]: The above is a different import group

**Comment by pmm-sumo** [comment]
README.md and sample config + config test needs to be updated as well

File: processor/k8sprocessor/client_test.go [hunk 2]
```diff
@@ -41,25 +42,26 @@ func selectors() (labels.Selector, fields.Selector) {
...
+		Associations: associations,
+		Informer:     kube.NewFakeInformer(cs, "", ls, fs),
+		StopCh:       make(chan struct{}),
 	}, nil
 }
 
-// GetPodByIP looks up FakeClient.Pods map by the provided string.
-func (f *fakeClient) GetPodByIP(ip string) (*kube.Pod, bool) {
-	p, ok := f.Pods[ip]
+// GetPod looks up FakeClient.Pods map by the provided string.
```
Comment (pmm-sumo) [inline processor/k8sprocessor/client_test.go:62]:
```suggestion
// GetPod looks up FakeClient.Pods map by the provided string,
// which might represent either IP address or Por UID.
```

File: processor/k8sprocessor/doc.go [hunk 3]
```diff
@@ -16,12 +16,26 @@
 //
 // The processor automatically discovers k8s resources (pods), extracts metadata from them and adds the
 // extracted metadata to the relevant spans, metrics and logs. The processor use the kubernetes API to discover all pods
-// running in a cluster, keeps a record of their IP addresses and interesting metadata. Upon receiving telemetry data,
+// running in a cluster, keeps a record of their IP addresses, pod UIDs and interesting metadata. Upon receiving telemetry data,
```
Comment (pmm-sumo) [inline processor/k8sprocessor/doc.go:19]: I would restructure the whole section starting from "Upon receiving telemetry data,..."

Like:

The rules for associating the source record with specific Pod Metadata are configured via `pod_association` key. It represents a list of rules that are executed in order until the first one is able to do the match.  Each rule is specified as a pair of `from` (representing the rule type) and `name` (representing the extracted key name). Following rule types are available:

* `from: labels` - takes the attribute from resource attributes (the value can contain either IP address or Pod UID)
* `from: connection` - takes the IP attribute from connection context (if available) and automatically associates it with `k8s.pod.ip` attribute

File: processor/k8sprocessor/kube/client.go [hunk 4]
```diff
@@ -279,15 +267,30 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
...
 	}
-	c.Pods[pod.Status.PodIP] = newPod
+	if pod.Status.PodIP != "" {
+		// compare initial scheduled timestamp for existing pod and new pod with same IP
+		// and only replace old pod if scheduled time of new pod is newer? This should fix
+		// the case where scheduler has assigned the same IP to a new pod but update event for
+		// the old pod came in later
+		func() {
+			if p, ok := c.Pods[pod.Status.PodIP]; ok {
+				if p.StartTime != nil && pod.Status.StartTime.Before(p.StartTime) {
```
Comment (pmm-sumo) [inline processor/k8sprocessor/kube/client.go:277]: I think the same applies for PodUID, the flow needs to be modified a bit. BTW, I presume `pod.UID` is never `nil` nor empty?

Comment (pmatyjasek-sumo) [reply processor/k8sprocessor/kube/client.go:277]: I'm not sure if this should be changed. There is no possibility to get the same UID for different pods

File: processor/k8sprocessor/pod_association.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	"go.opentelemetry.io/collector/client"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+)
+
+func k8sPodAssociationFromAttributes(ctx context.Context, attrs pdata.AttributeMap, associations kube.Associations) map[string]string {
+	var pod_ip, client_ip, context_ip string
+	podAssociation := make(map[string]string)
+
+	pod_ip = stringAttributeFromMap(attrs, k8sIPLabelName)
```
Comment (pmm-sumo) [inline processor/k8sprocessor/pod_association.go:32]: I think this and the below should be associated within specific association name sections

File: processor/k8sprocessor/kube/client.go (hunk 4 above)
Comment (pmm-sumo) [reply processor/k8sprocessor/kube/client.go:277]: Good point, perhaps worth including as a comment? :)

File: processor/k8sprocessor/pod_association.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+				continue
+			}
+			if asso.Name == "ip" && asso.From == "label" && client_ip != "" {
+				podAssociation[k8sIPLabelName] = client_ip
+				continue
+			}
+			if asso.Name == "ip" && asso.From == "connection" && context_ip != "" {
+				podAssociation[k8sIPLabelName] = context_ip
+				continue
+			}
```
Comment (sumo-drosiek) [inline processor/k8sprocessor/pod_association.go:57]: I would change that construction to `switch asso.From`. Will be more readable imho

Also don't we want check `k8s.pod.ip` for the `label` only

File: processor/k8sprocessor/pod_association.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	return podAssociation
+}
+
+func stringAttributeFromMap(attrs pdata.AttributeMap, key string) string {
+	if val, ok := attrs.Get(key); ok {
+		if val.Type() == pdata.AttributeValueSTRING {
+			return val.StringVal()
+		}
+	}
+	return ""
```
Comment (sumo-drosiek) [inline processor/k8sprocessor/pod_association.go:77]: I believe we can use `tracetranslator.AttributeValueToString` here

https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/protospan_translation.go#L113-L147

File: processor/k8sprocessor/options.go [hunk 8]
```diff
@@ -234,3 +234,17 @@ func WithFilterFields(filters ...FieldFilterConfig) Option {
 		return nil
 	}
 }
+
+func WithExtractPodAssociations(podAssociations ...PodAssociationConfig) Option {
+	return func(p *kubernetesprocessor) error {
+		associations := []kube.Association{}
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/options.go:240]: Preallocate?

```suggestion
		associations := make([]kube.Association, 0, len(podAssociations))
```

Btw why not directly act on `p.podAssociations.Associations`?

File: processor/k8sprocessor/options.go [hunk 9]
```diff
@@ -234,3 +234,17 @@ func WithFilterFields(filters ...FieldFilterConfig) Option {
 		return nil
 	}
 }
+
+func WithExtractPodAssociations(podAssociations ...PodAssociationConfig) Option {
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/options.go:238]: Comment?

File: processor/k8sprocessor/config.go [hunk 10]
```diff
@@ -39,6 +39,8 @@ type Config struct {
 	// Filter section allows specifying filters to filter
 	// pods by labels, fields, namespaces, nodes, etc.
 	Filter FilterConfig `mapstructure:"filter"`
+
+	Association []PodAssociationConfig `mapstructure:"pod_association"`
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/config.go:43]: Describe this with a comment?

File: processor/k8sprocessor/doc.go [hunk 11]
```diff
@@ -16,12 +16,26 @@
 //
 // The processor automatically discovers k8s resources (pods), extracts metadata from them and adds the
 // extracted metadata to the relevant spans, metrics and logs. The processor use the kubernetes API to discover all pods
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/doc.go:18]:
```suggestion
// extracted metadata to the relevant spans, metrics and logs. The processor uses the kubernetes API to discover all pods
```

File: processor/k8sprocessor/kube/client.go [hunk 12]
```diff
@@ -45,18 +45,19 @@ type WatchClient struct {
...
 
 // Extract deployment name from the pod name. Pod name is created using
 // format: [deployment-name]-[Random-String-For-ReplicaSet]-[Random-String-For-Pod]
 var dRegex = regexp.MustCompile(`^(.*)-[0-9a-zA-Z]*-[0-9a-zA-Z]*$`)
 
 // New initializes a new k8s Client.
-func New(logger *zap.Logger, apiCfg k8sconfig.APIConfig, rules ExtractionRules, filters Filters, newClientSet APIClientsetProvider, newInformer InformerProvider) (Client, error) {
-	c := &WatchClient{logger: logger, Rules: rules, Filters: filters, deploymentRegex: dRegex, stopCh: make(chan struct{})}
+func New(logger *zap.Logger, apiCfg k8sconfig.APIConfig, rules ExtractionRules, filters Filters, associations Associations, newClientSet APIClientsetProvider, newInformer InformerProvider) (Client, error) {
+	c := &WatchClient{logger: logger, Rules: rules, Filters: filters, Associations: associations, deploymentRegex: dRegex, stopCh: make(chan struct{})}
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/kube/client.go:60]: nitpick:

```suggestion
	c := &WatchClient{
		logger:          logger,
		Rules:           rules,
		Filters:         filters,
		Associations:    associations,
		deploymentRegex: dRegex,
		stopCh:          make(chan struct{}),
	}
```

File: processor/k8sprocessor/kube/kube.go [hunk 13]
```diff
@@ -125,3 +126,14 @@ type FieldExtractionRule struct {
 	// Full value is extracted when no regexp is provided.
 	Regex *regexp.Regexp
 }
+
+// Associations represents list of rules for associate Pod metadata with resource
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/kube/kube.go:130]: nitpick:

```suggestion
// Associations represent a list of rules for Pod metadata associations with resources
```

File: processor/k8sprocessor/options.go [hunk 14]
```diff
@@ -234,3 +234,18 @@ func WithFilterFields(filters ...FieldFilterConfig) Option {
 		return nil
 	}
 }
+
+// WithExtractPodAssoiciations allows specyfying options to associate pod metadata with incoming resource
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/options.go:238]:
```suggestion
// WithExtractPodAssociations allows specifying options to associate pod metadata with incoming resource
```

File: processor/k8sprocessor/processor.go [hunk 15]
```diff
@@ -107,56 +110,33 @@ func (kp *kubernetesprocessor) ProcessLogs(ctx context.Context, ld pdata.Logs) (
...
+	attrsToAdd := make(map[string]string)
+
+	for k, v := range podAttributes {
+		if !kp.passthroughMode {
+			attrsToAdd = kp.getAttributesForPod(v)
+			for key, val := range attrsToAdd {
+				resource.Attributes().InsertString(key, val)
+			}
+		}
+		resource.Attributes().InsertString(k, v)
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/processor.go:134]:
```suggestion
	for k, v := range podAttributes {
		if !kp.passthroughMode {
			attrsToAdd := kp.getAttributesForPod(v)
			for key, val := range attrsToAdd {
				resource.Attributes().InsertString(key, val)
			}
		}
		resource.Attributes().InsertString(k, v)
```

File: processor/k8sprocessor/doc.go [hunk 16]
```diff
@@ -16,12 +16,26 @@
...
-// that sent the telemetry data.
-// If a match is found, the cached metadata is added to the data as resource attributes.
+// running in a cluster, keeps a record of their IP addresses, pod UIDs and interesting metadata.
+// The rules for associating the source record with specific Pod Metadata are configured via "pod_association" key.
+// It represents a list of rules that are executed in order until the first one is able to do the match.
+// Each rule is specified as a pair of from (representing the rule type) and name (representing the extracted key name).
+// Following rule types are available:
+// 	  from: "labels" - takes the attribute from resource attributes (the value can contain either IP address or Pod UID)
+// 	  from: "connection" - takes the IP attribute from connection context (if available) and automatically
+// 						   associates it with "k8s.pod.ip" attribute
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/doc.go:26]: Mixed tabs and spaces:

```suggestion
// Following rule types are available:
//   from: "labels" - takes the attribute from resource attributes (the value can contain either IP address or Pod UID)
//   from: "connection" - takes the IP attribute from connection context (if available) and automatically
//     associates it with "k8s.pod.ip" attribute
```

**Review by pmalek-sumo — APPROVED** [review]

File: processor/k8sprocessor/config.go [hunk 17]
```diff
@@ -176,7 +176,14 @@ type FieldFilterConfig struct {
 	Op string `mapstructure:"op"`
 }
 
+// PodAssociationConfig allows specyfying rules for associate resources
```
Comment (pmalek-sumo) [inline processor/k8sprocessor/config.go:179]:
```suggestion
// PodAssociationConfig allows specifying rules for associating resources
```

**Pull request closed without merging**